File: core/scanner/scanner_test.py
"""Scanner模块的Python测试（使用unittest）"""

import unittest
import asyncio
from scanner import Scanner, MockScanner


class TestScanner(unittest.TestCase):
    def test_new_scanner(self):
        """测试创建Scanner实例"""
        s = Scanner()
        self.assertIsNotNone(s)
    
    def test_scanner_scan(self):
        """测试Scan方法"""
        s = Scanner()
        agents = s.scan()
        self.assertIsInstance(agents, list)
        self.assertEqual(len(agents), 0)  # 初始状态应该为空
    
    def test_scanner_watch(self):
        """测试Watch方法"""
        s = Scanner()
        
        async def test_watch():
            events = []
            # 只运行0.3秒，避免无限循环
            async for event in s.watch(timeout=0.1):
                events.append(event)
                if len(events) >= 2:  # 最多获取2个事件
                    break
            return events
        
        events = asyncio.run(test_watch())
        # 初始状态应该至少有一个事件（添加mock agent）
        self.assertGreaterEqual(len(events), 1)
    
    def test_mock_scanner(self):
        """测试MockScanner"""
        mock = MockScanner(
            agents=[
                {
                    "id": "test-agent-1",
                    "name": "Test Agent",
                    "type": "test",
                    "status": "running",
                    "pid": 12345,
                }
            ]
        )
        
        agents = mock.scan()
        self.assertEqual(len(agents), 1)
        self.assertEqual(agents[0]["id"], "test-agent-1")
    
    def test_scanner_add_agent(self):
        """测试添加Agent"""
        s = Scanner()
        agent = {
            "id": "test-agent-2",
            "name": "Test Agent 2",
            "type": "test",
            "status": "running",
            "pid": 12346,
        }
        s.add_agent(agent)
        
        agents = s.scan()
        self.assertEqual(len(agents), 1)
        self.assertEqual(agents[0]["id"], "test-agent-2")
    
    def test_scanner_remove_agent(self):
        """测试移除Agent"""
        s = Scanner()
        agent = {
            "id": "test-agent-3",
            "name": "Test Agent 3",
            "type": "test",
            "status": "running",
            "pid": 12347,
        }
        s.add_agent(agent)
        s.remove_agent("test-agent-3")
        
        agents = s.scan()
        self.assertEqual(len(agents), 0)


if __name__ == "__main__":
    unittest.main()

File: core/scanner/scanner_test.go
package scanner

import (
	"context"
	"testing"
	"time"

	"github.com/agent-toolkit/core"
)

func TestNewScanner(t *testing.T) {
	s := NewScanner()
	if s == nil {
		t.Fatal("NewScanner() returned nil")
	}
}

func TestScanner_Scan(t *testing.T) {
	s := NewScanner()
	agents, err := s.Scan()
	if err != nil {
		t.Fatalf("Scan() error: %v", err)
	}

	// 扫描结果应该是空的，因为没有运行的Agent
	if len(agents) != 0 {
		t.Errorf("Scan() returned %d agents, want 0", len(agents))
	}
}

func TestScanner_Watch(t *testing.T) {
	s := NewScanner()
	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
	defer cancel()

	events, err := s.Watch(ctx)
	if err != nil {
		t.Fatalf("Watch() error: %v", err)
	}

	// 应该没有事件
	select {
	case event := <-events:
		t.Errorf("Watch() received event: %v", event)
	case <-ctx.Done():
		// 正常超时
	}
}

func TestScanner_ScanWithMock(t *testing.T) {
	// 创建一个模拟的Scanner
	mock := &MockScanner{
		agents: []core.Agent{
			{
				ID:     "test-agent-1",
				Name:   "Test Agent",
				Type:   "test",
				Status: "running",
				PID:    12345,
			},
		},
	}

	agents, err := mock.Scan()
	if err != nil {
		t.Fatalf("Mock Scan() error: %v", err)
	}

	if len(agents) != 1 {
		t.Fatalf("Mock Scan() returned %d agents, want 1", len(agents))
	}

	if agents[0].ID != "test-agent-1" {
		t.Errorf("Agent ID = %s, want test-agent-1", agents[0].ID)
	}
}

// MockScanner 是一个模拟的Scanner实现
type MockScanner struct {
	agents []core.Agent
}

func (m *MockScanner) Scan() ([]core.Agent, error) {
	return m.agents, nil
}

func (m *MockScanner) Watch(ctx context.Context) (<-chan core.AgentEvent, error) {
	ch := make(chan core.AgentEvent)
	go func() {
		defer close(ch)
		<-ctx.Done()
	}()
	return ch, nil
}